### PR TITLE
MULE-9487: Extension support for TopLevel elements import

### PR DESCRIPTION
**Pull request by alegmarra**

### CODE REVIEW

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/describer/MuleExtensionAnnotationParser.java [hunk 1]
```diff
@@ -127,6 +129,31 @@ public static Extension getExtension(Class<?> extensionType)
...
 
+    static <T extends Annotation> List<T> parseRepeatableAnnotation(Class<?> extensionType, Class<T> annotation,
+                                                                    Function<Annotation, T[]> containerConsumer)
+    {
+        List<T> annotationDeclarations = Collections.emptyList();
+
+        Repeatable repeatableContainer = annotation.getAnnotation(Repeatable.class);
+        if (repeatableContainer != null)
+        {
+            //TODO FIXME look parent objects
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/describer/MuleExtensionAnnotationParser.java]: This has to be updated once https://github.com/mulesoft/mule/pull/2554 is merged, in order to rehuse the annotation lookup when inheritance is used

File: modules/cxf/src/test/java/org/mule/module/cxf/ProxyWSDLRewriteSchemaLocationsTestCase.java [hunk 2]
```diff
@@ -106,7 +106,7 @@ private Document getWsdl(MuleMessage response) throws Exception
 
     private String getLocation(Element schemaImport)
     {
-        return schemaImport.getAttributes().getNamedItem("schemaLocation").getNodeValue();
+        return schemaImport.getAttributes().getNamedItem("namespaceLocation").getNodeValue();
```
Comment (marianogonzalez) [inline modules/cxf/src/test/java/org/mule/module/cxf/ProxyWSDLRewriteSchemaLocationsTestCase.java]: revert this change.

File: modules/extensions-spring-support/src/main/java/org/mule/module/extension/internal/capability/xml/XmlModelEnricher.java [hunk 3]
```diff
@@ -30,23 +32,25 @@
...
     {
         Xml xml = extractAnnotation(describingContext.getExtensionDeclarer().getExtensionDeclaration(), Xml.class);
         ExtensionDeclarer descriptor = describingContext.getExtensionDeclarer();
-        descriptor.withModelProperty(createXmlModelProperty(xml, descriptor));
+        ExtensionDeclaration extensionDeclaration = descriptor.getExtensionDeclaration();
+        descriptor.withModelProperty(createXmlModelProperty(xml, extensionDeclaration.getName(), extensionDeclaration.getVersion()));
     }
 
-    private XmlModelProperty createXmlModelProperty(Xml xml, ExtensionDeclarer descriptor)
+    public XmlModelProperty createXmlModelProperty(Xml xml, String extensionName, String extensionVersion)
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/module/extension/internal/capability/xml/XmlModelEnricher.java]: this shouldn't be public

File: modules/extensions-spring-support/src/main/java/org/mule/module/extension/internal/capability/xml/schema/builder/SchemaBuilder.java [hunk 4]
```diff
@@ -889,23 +907,38 @@ private boolean shouldGeneratePojoChildElements(Class<?> type)
...
+                       !subTypesMapping.getSubTypes(parameterModel.getType()).isEmpty() ||
                        (IntrospectionUtils.isInstantiable(type) && ExpressionSupport.REQUIRED != parameterModel.getExpressionSupport());
             }
         };
     }
 
-    private void registerPojoSubtypes(List<MetadataType> subTypes , ExplicitGroup all)
+    private void addImportedTypeRef(Class<?> extensionType, ParameterModel parameterModel, ExplicitGroup all)
+    {
+        XmlModelProperty xml = new XmlModelEnricher().createXmlModelProperty(
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/module/extension/internal/capability/xml/schema/builder/SchemaBuilder.java]: if you need to reuse logic extract to a utils... do not put a ModelEnricher in here

File: modules/extensions-spring-support/src/main/java/org/mule/module/extension/internal/capability/xml/schema/builder/SchemaBuilder.java [hunk 5]
```diff
@@ -953,12 +986,16 @@ private void addAttributeAndElement(ExtensionType extensionType, ExplicitGroup a
                                                                          false,
                                                                          ExpressionSupport.NOT_SUPPORTED));
 
+        all.getParticle().add(objectFactory.createElement(createRefElement(elementRef, false)));
+    }
+
+    TopLevelElement createRefElement(QName elementRef, boolean isRequired)
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/module/extension/internal/capability/xml/schema/builder/SchemaBuilder.java]: can this be private?

File: modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/ImplicitConfigTestCase.java [hunk 6]
```diff
@@ -64,7 +64,7 @@ public void getImplicitConfig() throws Exception
 
     @Extension(name = "implicit")
     @Operations({ImplicitOperations.class})
-    @Xml(schemaLocation = "http://www.mulesoft.org/schema/mule/implicit", namespace = "implicit", schemaVersion = "1.0")
+    @Xml(namespaceLocation = "http://www.mulesoft.org/schema/mule/implicit", namespace = "implicit", schemaVersion = "1.0")
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/ImplicitConfigTestCase.java]: what's the need for this rename? Name matches devkit

File: modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/capability/xml/XmlModelEnricherTestCase.java [hunk 7]
```diff
@@ -62,7 +69,11 @@ public void enrichWithDefaultValues()
         assertThat(xmlProperty, is(notNullValue()));
         assertThat(xmlProperty.getSchemaVersion(), is(EXTENSION_VERSION));
         assertThat(xmlProperty.getNamespace(), is(EXTENSION_HYPHENAZED_NAME));
-        assertThat(xmlProperty.getSchemaLocation(), equalTo(String.format(DEFAULT_SCHEMA_LOCATION_MASK, EXTENSION_HYPHENAZED_NAME)));
+        assertThat(xmlProperty.getNamespaceUri(), equalTo(String.format(DEFAULT_NAMESPACE_LOCATION_MASK, EXTENSION_HYPHENAZED_NAME)));
+        assertThat(xmlProperty.getXsdFileName(), is(String.format(XSD_FILENAME_MASK, EXTENSION_HYPHENAZED_NAME)));
+        assertThat(xmlProperty.getSchemaLocation(), is(String.format(DEFAULT_SCHEMA_LOCATION_MASK,
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/capability/xml/XmlModelEnricherTestCase.java]: why all these changes?

File: modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/capability/xml/XmlModelEnricherTestCase.java [hunk 8]
```diff
@@ -74,7 +85,11 @@ public void enrichWithCustomNamespaceValue()
         assertThat(xmlProperty, is(notNullValue()));
         assertThat(xmlProperty.getSchemaVersion(), is(EXTENSION_VERSION));
         assertThat(xmlProperty.getNamespace(), is(NAMESPACE));
-        assertThat(xmlProperty.getSchemaLocation(), equalTo(String.format(DEFAULT_SCHEMA_LOCATION_MASK, NAMESPACE)));
+        assertThat(xmlProperty.getNamespaceUri(), equalTo(String.format(DEFAULT_NAMESPACE_LOCATION_MASK, NAMESPACE)));
+        assertThat(xmlProperty.getXsdFileName(), is(String.format(XSD_FILENAME_MASK, NAMESPACE)));
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/capability/xml/XmlModelEnricherTestCase.java]: idem

File: modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/capability/xml/XmlModelEnricherTestCase.java [hunk 9]
```diff
@@ -86,7 +101,11 @@ public void enrichWithCustomSchemaLocationValue()
         assertThat(xmlProperty, is(notNullValue()));
         assertThat(xmlProperty.getSchemaVersion(), is(EXTENSION_VERSION));
         assertThat(xmlProperty.getNamespace(), is(EXTENSION.toLowerCase()));
-        assertThat(xmlProperty.getSchemaLocation(), equalTo(SCHEMA_LOCATION));
+        assertThat(xmlProperty.getNamespaceUri(), equalTo(NAMESPACE_LOCATION));
+        assertThat(xmlProperty.getXsdFileName(), is(String.format(XSD_FILENAME_MASK, hyphenize(EXTENSION))));
+        assertThat(xmlProperty.getSchemaLocation(), is(String.format(DEFAULT_SCHEMA_LOCATION_MASK,
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/capability/xml/XmlModelEnricherTestCase.java]: idem

File: modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/resources/SpringBundleResourceContributorTestCase.java [hunk 10]
```diff
@@ -61,9 +65,10 @@
     @Before
     public void before()
     {
-        xmlModelProperty = new XmlModelProperty(EXTENSION_VERSION, EXTENSION_NAME, UNESCAPED_LOCATION_PREFIX + SCHEMA_LOCATION);
+        xmlModelProperty = new XmlModelProperty(EXTENSION_VERSION, EXTENSION_NAME, UNESCAPED_LOCATION_PREFIX + SCHEMA_LOCATION, SCHEMA_NAME, String.format("%s/%s/%s", UNESCAPED_LOCATION_PREFIX + SCHEMA_LOCATION, SchemaConstants.CURRENT_VERSION, SCHEMA_NAME));
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/resources/SpringBundleResourceContributorTestCase.java]: too long a line... Still don't understand what you're doing here though

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/describer/MuleExtensionAnnotationParser.java [hunk 11]
```diff
@@ -127,6 +129,31 @@ public static Extension getExtension(Class<?> extensionType)
         return parsedParameters;
     }
 
+    static <T extends Annotation> List<T> parseRepeatableAnnotation(Class<?> extensionType, Class<T> annotation,
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/describer/MuleExtensionAnnotationParser.java]: also apply this method on the sub types thingy

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/describer/MuleExtensionAnnotationParser.java [hunk 12]
```diff
@@ -127,6 +129,31 @@ public static Extension getExtension(Class<?> extensionType)
...
+        List<T> annotationDeclarations = Collections.emptyList();
+
+        Repeatable repeatableContainer = annotation.getAnnotation(Repeatable.class);
+        if (repeatableContainer != null)
+        {
+            //TODO FIXME look parent objects
+            Annotation container = extensionType.getAnnotation(repeatableContainer.value());
+            if (container != null)
+            {
+                annotationDeclarations = Arrays.asList(containerConsumer.apply(container));
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/describer/MuleExtensionAnnotationParser.java]: immutable

File: modules/extensions-support/src/test/java/org/mule/module/extension/vegan/ApplesBasket.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.module.extension.vegan;
+
+import org.mule.tck.testmodels.fruit.Apple;
+
+import java.util.List;
+
+public class ApplesBasket
```
Comment (marianogonzalez) [inline modules/extensions-support/src/test/java/org/mule/module/extension/vegan/ApplesBasket.java]: this should be defined in the same project as the Apple

File: modules/extensions-spring-support/src/main/java/org/mule/module/extension/internal/capability/xml/schema/builder/SchemaBuilder.java (hunk 5 above)
Comment (alegmarra) [inline modules/extensions-spring-support/src/main/java/org/mule/module/extension/internal/capability/xml/schema/builder/SchemaBuilder.java]: No, It's being used also by the ConnectionProviderSchemaDelegate

File: modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/ImplicitConfigTestCase.java (hunk 6 above)
Comment (alegmarra) [inline modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/ImplicitConfigTestCase.java]: If we use the schema imports as a rule, then we have:

```<xs:import 
namespace="http://www.mulesoft.org/schema/mule/implicit" schemaLocation="http://www.mulesoft.org/schema/mule/extension/current/mule-implicit.xsd"></xs:import>```

So changing that URI to `namespaceLocation` is less confusing

File: modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/capability/xml/XmlModelEnricherTestCase.java (hunk 7 above)
Comment (alegmarra) [inline modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/capability/xml/XmlModelEnricherTestCase.java]: testing the parsing of new fields available in the `XmlModelProperty`

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/SchemaBuilder.java [hunk 14]
```diff
@@ -158,6 +167,9 @@ public static SchemaBuilder newSchema(ExtensionModel extensionModel, String targ
         Optional<SubTypesModelProperty> subTypesProperty = extensionModel.getModelProperty(SubTypesModelProperty.class);
         builder.withTypeMapping(subTypesProperty.isPresent() ? subTypesProperty.get().getSubTypesMapping() : Collections.emptyMap());
 
+        Optional<Map<MetadataType, MetadataType>> importedTypes = extensionModel.getModelProperty(ImportedTypesModelProperty.class).map(ImportedTypesModelProperty::getImportedTypes);
+        builder.withImportedTypes(importedTypes.isPresent() ? importedTypes.get() : Collections.emptyMap());
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/SchemaBuilder.java]: guava

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/SchemaBuilder.java [hunk 15]
```diff
@@ -874,6 +888,12 @@ public void visitObject(ObjectType objectType)
 
                 defaultVisit(objectType);
 
+                if (importedTypes.get(objectType) != null)
+                {
+                    addImportedTypeRef(JavaTypeUtils.getType(importedTypes.get(objectType)), parameterModel, all);
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/SchemaBuilder.java]: static import

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/introspection/describer/MuleExtensionAnnotationParser.java [hunk 16]
```diff
@@ -127,6 +130,31 @@ public static Extension getExtension(Class<?> extensionType)
...
+    static <T extends Annotation> List<T> parseRepeatableAnnotation(Class<?> extensionType, Class<T> annotation,
+                                                                    Function<Annotation, T[]> containerConsumer)
+    {
+        List<T> annotationDeclarations = Collections.emptyList();
+
+        Repeatable repeatableContainer = annotation.getAnnotation(Repeatable.class);
+        if (repeatableContainer != null)
+        {
+            //TODO FIXME look parent objects
+            Annotation container = extensionType.getAnnotation(repeatableContainer.value());
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/introspection/describer/MuleExtensionAnnotationParser.java]: you didn't fix this one
